src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,885 @@
+        export const CannotExtractRangeContainingConditionalBreakOrContinueStatements: DiagnosticMessage = m("Cannot extract range containing conditional break or continue statements.");
+        export const CannotExtractRangeContainingConditionalReturnStatement: DiagnosticMessage = m("Cannot extract range containing conditional return statement.");
+        export const CannotExtractRangeContainingLabeledBreakOrContinueStatementWithTargetOutsideOfTheRange: DiagnosticMessage = m("Cannot extract range containing labeled break or continue with target outside of the range.");
+        export const CannotExtractRangeThatContainsWritesToReferencesLocatedOutsideOfTheTargetRangeInGenerators: DiagnosticMessage = m("Cannot extract range containing writes to references located outside of the target range in generators.");
+        export const TypeWillNotBeVisibleInTheNewScope = m("Type will not visible in the new scope.");
+        export const FunctionWillNotBeVisibleInTheNewScope = m("Function will not visible in the new scope.");
+        export const InsufficientSelection = m("Select more than a single identifier.");
+    }
+
+    export enum RangeFacts {
+        None = 0,
+        HasReturn = 1 << 0,
+        IsGenerator = 1 << 1,
+        IsAsyncFunction = 1 << 2,
+        UsesThis = 1 << 3
+    }
+
+    /**
+     * Represents an expression or a list of statements that should be extracted with some extra information
+     */
+    export interface TargetRange {
+        readonly range: Expression | Statement[];
+        readonly facts: RangeFacts;
+    }
+
+    /**
+     * Result of 'getRangeToExtract' operation: contains either a range or a list of errors
+     */
+    export interface RangeToExtract {
+        readonly targetRange?: TargetRange;
+        readonly errors?: Diagnostic[];
+    }
+
+    /*
+     * Scopes that can store newly extracted method
+     */
+    export type Scope = FunctionLikeDeclaration | SourceFile | ModuleBlock | ClassLikeDeclaration;
+
+
+    /**
+     * Result of 'extractRange' operation for a specific scope.
+     * Stores either a list of changes that should be applied to extract a range or a list of errors
+     */
+    export interface ExtractResultForScope {
+        readonly scope: Scope;
+        readonly scopeDescription: string;
+        readonly changes?: FileTextChanges[];
+        readonly errors?: Diagnostic[];
+    }
+
+    /**
+     * getRangeToExtract takes a span inside a text file and returns either an expression or an array
+     * of statements representing the minimum set of nodes needed to extract the entire span. This
+     * process may fail, in which case a set of errors is returned instead (these are currently
+     * not shown to the user, but can be used by us diagnostically)
+     */
+    export function getRangeToExtract(sourceFile: SourceFile, span: TextSpan): RangeToExtract {
+        const length = span.length || 0;

```

Comment:
This is more defensive than anything (since we don't have SNC on I can't be definitive about whether `length` is always `undefined` when the selection is empty or not). `undefined || 0` is `0` and e.g. `undefined + 1` is `NaN` so this ensures we don't have problems if `length` comes in as `undefined`